CHANGELOG.md

```diff
@@ -4,7 +4,9 @@
 
 ### Grafana Mimir
 
+* [CHANGE]: Querier: returning 422 when query hits `max_fetched_chunks_per_query` and `max_fetched_series_per_query` limits in the store-gateway. #4056
```

Comment:
How was tested the querier now returns 422? Have you done any manual test? We may consider an integration test (I can give you guidance).

I'm asking because the fact that we return 422 via gRPC from store-gateway doesn't mean the querier will honor it. We have to check the whole path and having an integration test may be nice.

Reply:
@pracucci As agreed, we will check the outcome of manual tests tomorrow. However, I run the existing integration test [`querier_test.TestQueryLimitsWithBlocksStorageRunningInMicroServices`](https://github.com/grafana/mimir/blob/9c82bb89c4703ede2ad158941688f4252ac407b1/integration/querier_test.go#L831), and realized that the error returned by the call to [`QueryRange()`](https://github.com/grafana/mimir/blob/9c82bb89c4703ede2ad158941688f4252ac407b1/integration/querier_test.go#L898) **does NOT** contain the required status code `422`.

However, I did some debugging and came to the following conclusion: the call to [`QueryRange()`](https://github.com/grafana/mimir/blob/9c82bb89c4703ede2ad158941688f4252ac407b1/integration/querier_test.go#L898) gets propagated to [`httpAPI.QueryRange()`](https://github.com/grafana/mimir/blob/9c82bb89c4703ede2ad158941688f4252ac407b1/vendor/github.com/prometheus/client_golang/api/prometheus/v1/api.go#L863), which calls [`apiClientImpl.DoGetFallback()`](https://github.com/grafana/mimir/blob/9c82bb89c4703ede2ad158941688f4252ac407b1/vendor/github.com/prometheus/client_golang/api/prometheus/v1/api.go#L882). The latter successfully returns `(*http.Response, []byte, Warnings, error)`, which first output **DOES contain** the required status code `422`, but that output gets ignored by [`httpAPI.QueryRange()`](https://github.com/grafana/mimir/blob/9c82bb89c4703ede2ad158941688f4252ac407b1/vendor/github.com/prometheus/client_golang/api/prometheus/v1/api.go#L882). I don't know whether `httpAPI` is used only for integration test purposes or it is actually used in production. In the first case, we would need to fix the integration test. In the second case, `httpAPI` should be fixed. What do you think @pracucci @dimitarvdimitrov?

Reply:
As i understand the problem is in the query client that we use in the integration tests. So the HTTP 422 gets from the store-gateway to the querier and from the querier to the HTTP response, but the golang client filters it out at some point. 

In production we don't run HTTP requests against the querier. At least not directly. The querier receives HTTP requests over gRPC from the query-scheduler, executes the queries and returns the results to a query-frontend (address of QF is in the original request).

I believe that these are the two HTTP handlers that handle a query in the querier - [query_range](https://github.com/grafana/mimir/blob/19cff491b4b27fd024aa4c6db646767b1158f743/vendor/github.com/prometheus/prometheus/web/api/v1/api.go#L464-L536), [instant query](https://github.com/grafana/mimir/blob/19cff491b4b27fd024aa4c6db646767b1158f743/vendor/github.com/prometheus/prometheus/web/api/v1/api.go#L396-L445). They both use some injected queriables that we set up in [NewqQurierHandler](https://github.com/grafana/mimir/blob/ca5cdff502c301cd1ced08b064c45c3a2685c58e/pkg/api/handlers.go#L236-L262)

They are HTTP handlers, but their HTTP response is sent over gRPC to the query-frontend. You can check the querier logic which receives HTTP requests over gRPC and send the responses to the query-frontend here [`frontendProcessor.runRequest`](https://github.com/grafana/mimir/blob/80a876193f386567e689f5b91d1a1439f4b340a1/pkg/querier/worker/frontend_processor.go#L140-L169)

So i believe that as those two HTTP handlers return the right HTTP codes, they will also get to the query-frontend and to the client (it looks like they do, but I haven't looked at the call chain between these handlers and the store-gateway client)

Reply:
Thank you @dimitarvdimitrov
I will then override the golang client in my integration test.
Could you please re-approve and merge my PR (I changed some names, as Marco recommended)?